traversal/procedure/GraphProcedure.java

```diff
@@ -247,81 +248,78 @@ private void registerEdge(StructureEdge<?, ?> structureEdge, boolean isForward)
             }
             if (from.order() > to.order()) return;
             ProcedureEdge<?, ?> edge = ProcedureEdge.of(from, to, structureEdge, isForward);
-            registerEdge(edge);
-        }
-
-        private void registerEdge(PlannerEdge.Directional<?, ?> plannerEdge) {
-            ProcedureVertex<?, ?> from = vertex(plannerEdge.from());
-            ProcedureVertex<?, ?> to = vertex(plannerEdge.to());
-            ProcedureEdge<?, ?> edge = ProcedureEdge.of(from, to, plannerEdge);
-            registerEdge(edge);
+            attachEdge(from, to, edge);
         }
 
-        public void registerEdge(ProcedureEdge<?, ?> edge) {
-            if (edge.from().equals(edge.to())) {
-                edge.from().loop(edge);
+        public void attachEdge(ProcedureVertex<?, ?> from, ProcedureVertex<?, ?> to, ProcedureEdge<?, ?> edge) {
```

Comment:
made the AP more explicit with `attachEdge(from, to, edge)`  which tells a better story of attaching the edge to the from/to vertex